### PR TITLE
Improvement/arsn 420 introduce put object hack

### PR DESCRIPTION
**Pull request by benzekrimaha**
Details on the implementation are within the commit description 

Issue : ARSN-420

### CODE REVIEW

File: lib/storage/metadata/mongoclient/MongoClientInterface.js [hunk 1]
```diff
@@ -899,35 +899,113 @@ class MongoClientInterface {
...
-            },
-        }, {
-            upsert: true,
-        }).then(() => cb()).catch((err) => {
-            log.error('putObjectNoVer: error putting obect with no versioning', { error: err.message });
-            return cb(errors.InternalError);
+    putObjectNoVer(collection, bucketName, objName, value, params, log, cb, originOp = 's3:LifecycleTransitionPut') {
+        const key = formatMasterKey(objName, params.vFormat);
+        const putFilter = { _id: key };
+        if (params && !params.doesNeedOplogUpdate) {
```
Comment (francoisferrand) [inline lib/storage/metadata/mongoclient/MongoClientInterface.js]: if no param, should stick to the "default" behavior
i.e.
```suggestion
        if (!params || !params.needOplogUpdate) {
```
or better yet
```suggestion
        if (!params?.needOplogUpdate) {
```

File: lib/storage/metadata/mongoclient/MongoClientInterface.js [hunk 2]
```diff
@@ -899,35 +899,113 @@ class MongoClientInterface {
...
-     * @return {undefined}
+     * Puts an object into a MongoDB collection. Depending on the parameters,
+     * the object is either directly put into the collection
+     * or the existing object is marked as deleted and a new object is inserted.
+     *
+     * @param {Object} collection - The MongoDB collection to put the object into.
+     * @param {string} bucketName - The name of the bucket the object belongs to.
+     * @param {string} objName - The name of the object.
+     * @param {Object} value - The value of the object.
+     * @param {Object} params - Additional parameters. If params.doesNotNeedOpogUpdate is true, the object is directly
```
Comment (francoisferrand) [inline lib/storage/metadata/mongoclient/MongoClientInterface.js]: should document the individual parameters:

```suggestion
     * @param {Object} params - Additional parameters
     * @param {bool} params.vFormat - ...
     * @param {bool} params.doesNotNeedOpogUpdate - fs true, the object is directly
```

File: lib/storage/metadata/mongoclient/MongoClientInterface.js (hunk 1 above)
Comment (francoisferrand) [reply lib/storage/metadata/mongoclient/MongoClientInterface.js]: somehow, the "structure" of the function looks a bit misleading to be: by having the negative condition, it makes it look like the "standard" case if to update log...

maybe worth reworking the overal organization of the change to make it more clear, something like:

```javascript
putObjectNoVer(...) {
    if (params.needOplogUpdate) {
        return putObjectNoVerWithOplogUpdate()
    }
    // [...] proceed with regular put object (old code)
}
```

File: lib/storage/metadata/mongoclient/MongoClientInterface.js [hunk 3]
```diff
@@ -899,35 +899,113 @@ class MongoClientInterface {
...
+        }, putFilter);
+
+        return async.waterfall([
+            // Adding delete flag when getting the object
+            // to avoid having race conditions.
+            next => collection.findOneAndUpdate(findFilter, {
+                $set: {
+                    '_id': key,
+                    'value.deleted': true,
+                },
```
Comment (francoisferrand) [inline lib/storage/metadata/mongoclient/MongoClientInterface.js]: this is actually `updateDeleteFilter`...

File: lib/storage/metadata/mongoclient/MongoClientInterface.js [hunk 4]
```diff
@@ -899,35 +899,113 @@ class MongoClientInterface {
...
-            $set: {
-                _id: masterKey,
-                value: objVal,
-            },
-        }, {
-            upsert: true,
-        }).then(() => cb()).catch((err) => {
-            log.error('putObjectNoVer: error putting obect with no versioning', { error: err.message });
-            return cb(errors.InternalError);
+    putObjectNoVer(collection, bucketName, objName, value, params, log, cb, originOp = 's3:LifecycleTransitionPut') {
```
Comment (francoisferrand) [inline lib/storage/metadata/mongoclient/MongoClientInterface.js]: why would it have such a default?
- this is an abstract metadata interface, there should not be any default related to a specific use case.
- in the scenario we are trying to solve, there is no "lifecycle" on this specific (user!) operation: this should be more something like `ReplaceArchivedObject`... 

* maybe just using `PutObject` would be enough to identify this case (e.g. `action=s3:PutObject` with `deleted=true`) without triggering unexpected event. This would allow getting the info cleanly from the object value, and precisely reflect what operation happened...
* if you need to introduce an extra `originOp`, it should come from cloudserver --> extra `params`

File: lib/storage/metadata/mongoclient/MongoClientInterface.js [hunk 5]
```diff
@@ -899,35 +899,113 @@ class MongoClientInterface {
...
+                {
+                    updateOne: {
+                        filter: putFilter,
+                        update: {
+                            $set: { _id: key, value },
+                        },
+                        upsert: true,
+                    },
+                },
+            ], { ordered: true }).then(() => next(null)).catch(err => next(err)),
```
Comment (francoisferrand) [inline lib/storage/metadata/mongoclient/MongoClientInterface.js]:
```suggestion
            ], { ordered: true }).then(() => next(null)).catch(next),
```

File: lib/storage/metadata/mongoclient/MongoClientInterface.js [hunk 6]
```diff
@@ -973,8 +1051,7 @@ class MongoClientInterface {
                 return putObjectVer(c, bucketName, objName, objVal, _params, log,
                     cb);
             }
-            return this.putObjectNoVer(c, bucketName, objName, objVal,
-                _params, log, cb);
+            return this.putObjectNoVer(c, bucketName, objName, objVal, _params, log, cb);
```
Comment (francoisferrand) [inline lib/storage/metadata/mongoclient/MongoClientInterface.js:1071]: no change here?

File: lib/storage/metadata/mongoclient/MongoClientInterface.js [hunk 7]
```diff
@@ -899,35 +899,132 @@ class MongoClientInterface {
...
+                value,
             },
         }, {
             upsert: true,
-        }).then(() => cb()).catch((err) => {
-            log.error('putObjectNoVer: error putting obect with no versioning', { error: err.message });
+        }).then(() => cb()).catch(err => {
+            log.error('internalPutObject: error putting object',
+                { bucket: bucketName, object: key, error: err.message });
             return cb(errors.InternalError);
```
Comment (francoisferrand) [inline lib/storage/metadata/mongoclient/MongoClientInterface.js]: probably no point changing this, old code works just as well ?
(and `putObjectNoVer: error putting obect with no versioning` error message was more appropriate, giving details about which part of the code fails...)

File: lib/storage/metadata/mongoclient/MongoClientInterface.js [hunk 8]
```diff
@@ -899,35 +899,132 @@ class MongoClientInterface {
...
+     * @param {Function} cb - The callback function to call when the operation is complete.
+     * It is called with an error if there is an issue with the operation.
+     * @returns {void}
+     */
+    putObjectNoVerWithOplogUpdate(collection, bucketName, objName, value, params, log, cb) {
+        const key = formatMasterKey(objName, params.vFormat);
+        const putFilter = { _id: key };
+        // filter used when finding and updating object
+        const findFilter = Object.assign({
+            _id: key,
```
Comment (francoisferrand) [inline lib/storage/metadata/mongoclient/MongoClientInterface.js]: redundant with `, putFilter` a few lines below...

also, whenever possible, prefer to use `...` to "unpack" object instead of Object.assign, as it will be easier to migrate to typescript


```suggestion
    const findFilter = {
        ...putFilter,
        $or: [
            { 'value.deleted': { $exists: false } },
            { 'value.deleted': { $eq: false } },
        ],
    };
```

File: lib/storage/metadata/mongoclient/MongoClientInterface.js [hunk 9]
```diff
@@ -899,35 +899,132 @@ class MongoClientInterface {
...
+        }, putFilter);
+        const updateDeleteFilter = Object.assign({
+            '_id': key,
+            'value.deleted': true,
+        }, putFilter);
+        return async.waterfall([
+            // Adding delete flag when getting the object
+            // to avoid having race conditions.
+            next => collection.findOneAndUpdate(findFilter, {
+                $set: updateDeleteFilter,
```
Comment (francoisferrand) [inline lib/storage/metadata/mongoclient/MongoClientInterface.js:976]: overall, I wonder if the introduction of all these filters actually helps with readability: because it reduces locality and decouples the function call, filter and "update". What do you think?

**Review by francoisferrand — APPROVED** [review]

File: lib/storage/metadata/mongoclient/MongoClientInterface.js [hunk 10]
```diff
@@ -899,35 +899,132 @@ class MongoClientInterface {
...
+            }, {
+                upsert: false,
+            }).then(doc => {
+                log.info('doc value', { doc: doc.value });
```
Comment (welansari) [inline lib/storage/metadata/mongoclient/MongoClientInterface.js]: Is this a leftover from debugging ?

File: lib/storage/metadata/mongoclient/MongoClientInterface.js [hunk 11]
```diff
@@ -899,35 +899,132 @@ class MongoClientInterface {
...
+            _id: key,
+            $or: [
+                { 'value.deleted': { $exists: false } },
+                { 'value.deleted': { $eq: false } },
+            ],
+        }, putFilter);
+        const updateDeleteFilter = Object.assign({
+            '_id': key,
+            'value.deleted': true,
+        }, putFilter);
```
Comment (welansari) [inline lib/storage/metadata/mongoclient/MongoClientInterface.js]: Not sure what's the point of `Object.assign` since we already set the `_id` key in the filters ?

File: lib/storage/metadata/mongoclient/MongoClientInterface.js (hunk 9 above)
Comment (benzekrimaha) [reply lib/storage/metadata/mongoclient/MongoClientInterface.js:976]: The fact that the filters are defined at the top may make it hard to follow indeed but it  reduces redundancy , and maintainability is easier than in the inline approach. We either make the code DRY (Don't Repeat Yourself) or easier to understand with the inline definition on each mongo op ( I don't really have reserves on both approaches this one just seemed cleaner to me)

File: lib/storage/metadata/mongoclient/MongoClientInterface.js (hunk 11 above)
Comment (benzekrimaha) [reply lib/storage/metadata/mongoclient/MongoClientInterface.js]: indeed , fixed here : https://github.com/scality/Arsenal/pull/2249/commits/043b64c3d17ad2bbcc3ea34ddc2b3f6590c436b1#diff-9c207befbd8f4d79a3b6e3d7d143444e804ca1684236b8fd9fa0f9e03c53477aR961-R971

File: lib/storage/metadata/mongoclient/MongoClientInterface.js (hunk 8 above)
Comment (benzekrimaha) [reply lib/storage/metadata/mongoclient/MongoClientInterface.js]: done here : https://github.com/scality/Arsenal/pull/2249/commits/043b64c3d17ad2bbcc3ea34ddc2b3f6590c436b1#diff-9c207befbd8f4d79a3b6e3d7d143444e804ca1684236b8fd9fa0f9e03c53477aR961-R971

File: lib/storage/metadata/mongoclient/MongoClientInterface.js (hunk 7 above)
Comment (benzekrimaha) [reply lib/storage/metadata/mongoclient/MongoClientInterface.js]: code was kept as is, the error message is fixed here : https://github.com/scality/Arsenal/pull/2249/commits/043b64c3d17ad2bbcc3ea34ddc2b3f6590c436b1#diff-9c207befbd8f4d79a3b6e3d7d143444e804ca1684236b8fd9fa0f9e03c53477aR935

**Review by ghost — APPROVED** [review]
